azure/services/securitygroups/securitygroups.go

```diff
@@ -80,15 +82,34 @@ func (s *Service) Reconcile(ctx context.Context) error {
 
 	var resErr error
 
+	// Retrieve the last applied security rules for all NSGs.
+	lastAppliedSecurityRulesAll, err := s.Scope.AnnotationJSON(azure.SecurityRuleLastAppliedAnnotation)
+	if err != nil {
+		log.V(4).Info("failed to retrieve last applied security rules annotation", "error", err)
+		resErr = err
+	}
+
 	// We go through the list of security groups to reconcile each one, independently of the result of the previous one.
 	// If multiple errors occur, we return the most pressing one.
 	//  Order of precedence (highest -> lowest) is: error that is not an operationNotDoneError (i.e. error creating) -> operationNotDoneError (i.e. creating in progress) -> no error (i.e. created)
-	for _, nsgSpec := range specs {
+	for _, resourceSpec := range specs {
+		nsgSpec := resourceSpec.(*NSGSpec)
+
 		if _, err := s.CreateOrUpdateResource(ctx, nsgSpec, serviceName); err != nil {
 			if !azure.IsOperationNotDoneError(err) || resErr == nil {
 				resErr = err
 			}
 		}
+
+		if len(nsgSpec.NewAnnotation) == 0 {
```

Comment:
So this is essentially relying on `CreateOrUpdateResource` to itself call `Parameters` which will modify the input parameter `nsgSpec` and set `nsgSpec.NewAnnotation`, is that correct?

If so it's a bit of a fragile contract, would it be better to just dedicate a new function `calculateNewAnnotation` that explicitly takes `nsgSpec` and returns the new annotation that should be set? The annotation should just be everything that's in spec I think?

Reply:
Yup you're right. This is definitely a hacky solution and I think it came from a misunderstanding that the updated annotation can be different from the spec. Will change this!